### PR TITLE
Push XR projection matrix generation down to platform implementations

### PR DESCRIPTION
**Pull request by Alex-MSFT**
This PR attempts to address #473, which effectively expanded the projection matrix to maintain the aspect ratio, but incorrectly calculated the real FOV.  I think this was occurring because we were applying both the angle FOV values with the aspect ratio from the color texture.

In this PR I'm pushing the generation of projection matrices down to the platform level implementations.  For ARCore and ARKit we are provided with valid projection matrices.  For OpenXR we need to generate them ourselves based off the FoV angles.

Here are some comparisons from my testing attempting to measure a 5' long measuring tape.  In both cases I measured by getting close to the tape and walking along its length then stepping back to a reasonable distance to take a picture.

old:

new:

### CODE REVIEW

File: Plugins/NativeXr/Source/NativeXr.cpp [hunk 1]
```diff
@@ -46,30 +46,19 @@ namespace
...
-        const float aspectRatio = static_cast<float>(view.ColorTextureSize.Width) / static_cast<float>(view.ColorTextureSize.Height);
-        const float deltax = (r - l);
-        const float deltay = (t - b);
-        const float afterTangentAspectRatio = deltax / deltay;
-        const float compensationRatio = afterTangentAspectRatio / aspectRatio;
-        const float tc{std::tanf(view.FieldOfView.AngleUp * compensationRatio) * n};
-        const float bc{std::tanf(view.FieldOfView.AngleDown * compensationRatio) * n};
+        // Calculate total vertical and horizontal FoV and the aspect ratio.
+        const float verticalFoV{abs(view.FieldOfView.AngleUp) + abs(view.FieldOfView.AngleDown)};
+        const float horizontalFoV{abs(view.FieldOfView.AngleRight) + abs(view.FieldOfView.AngleLeft)};
```
Comment (Alex-MSFT) [inline Plugins/NativeXr/Source/NativeXr.cpp:51]: This solution assumes symmetric FOVs.  Do we have cases in our use cases on OpenXR with asymmetric FOVs?

@syntheticmagus @rgerd @chrisfromwork

Comment (chrisfromwork) [reply Plugins/NativeXr/Source/NativeXr.cpp:51]: I'm not sure that we can assume FOVs are symmetrical. I can test this out on a HoloLens tomorrow

Comment (chrisfromwork) [reply Plugins/NativeXr/Source/NativeXr.cpp:51]: Was able to get to this today. HoloLens FOV's aren't symmetrical. we should sync on this offline tomorrow.

Comment (Alex-MSFT) [reply Plugins/NativeXr/Source/NativeXr.cpp:51]: Sounds good for now I am going to fork the behavior and keep the old implementation for asymmetric FoVs.  We should investigate if whether the old calculation was correct for these cases.

File: Plugins/NativeXr/Source/NativeXr.cpp [hunk 2]
```diff
@@ -74,6 +93,22 @@ namespace
         return bxResult;
     }
 
+    std::array<float, 16> CreateProjectionMatrix(const xr::System::Session::Frame::View& view)
+    {
+        const bool isCentered{
+            abs(view.FieldOfView.AngleUp + view.FieldOfView.AngleDown) < .01 &&
+            abs(view.FieldOfView.AngleRight + view.FieldOfView.AngleLeft) < .01};
```
Comment (chrisfromwork) [inline Plugins/NativeXr/Source/NativeXr.cpp:100]: nit: can we get 0.01 defined as a constexpr?

File: Plugins/NativeXr/Source/NativeXr.cpp (hunk 1 above)
Comment (Alex-MSFT) [reply Plugins/NativeXr/Source/NativeXr.cpp]: After a lengthy discussion, we're going to instead push projection matrix creation down to platform specific implementations.

File: Dependencies/xr/Source/ARCore/XR.cpp [hunk 3]
```diff
@@ -526,18 +526,16 @@ namespace xr
...
-                float b = projectionMatrix[1][1];
-
-                float aspectRatio = b / a;
-                float fieldOfView = std::atan(1.0f / b);
-
-                // Set the horizontal and vertical field of view
-                ActiveFrameViews[0].FieldOfView.AngleDown = -(ActiveFrameViews[0].FieldOfView.AngleUp = fieldOfView);
-                ActiveFrameViews[0].FieldOfView.AngleLeft = -(ActiveFrameViews[0].FieldOfView.AngleRight = fieldOfView * aspectRatio);
+                // Copy the projection matrix values into the view.
+                for(int row = 0; row < 4; row++)
```
Comment (CedricGuillemet) [inline Dependencies/xr/Source/ARCore/XR.cpp:532]: use memcpy?

Comment (Alex-MSFT) [reply Dependencies/xr/Source/ARCore/XR.cpp]: Good idea, done!

File: Dependencies/xr/Source/ARCore/XR.cpp [hunk 4]
```diff
@@ -526,18 +526,8 @@ namespace xr
...
-                float b = projectionMatrix[1][1];
-
-                float aspectRatio = b / a;
-                float fieldOfView = std::atan(1.0f / b);
-
-                // Set the horizontal and vertical field of view
-                ActiveFrameViews[0].FieldOfView.AngleDown = -(ActiveFrameViews[0].FieldOfView.AngleUp = fieldOfView);
-                ActiveFrameViews[0].FieldOfView.AngleLeft = -(ActiveFrameViews[0].FieldOfView.AngleRight = fieldOfView * aspectRatio);
+                ArCamera_getProjectionMatrix(session, camera, DepthNearZ, DepthFarZ, glm::value_ptr(projectionMatrix));
+                memcpy(ActiveFrameViews[0].ProjectionMatrix.data(), &projectionMatrix, sizeof(float) * 16);
```
Comment (ryantrem) [inline Dependencies/xr/Source/ARCore/XR.cpp:530]: Can't we statically get the size of the whole array? Seems like we shouldn't need to repeat the `16` here.

Comment (ryantrem) [inline Dependencies/xr/Source/ARCore/XR.cpp:530]: If you are doing a `memcpy` doesn't that indicate the memory layout is exactly the same? And if that is the case, can't we just cast `ProjectionMatrix` and pass it directly to `ArCamera_getProjectionMatrix`?

Comment (Alex-MSFT) [reply Dependencies/xr/Source/ARCore/XR.cpp:530]: Good point I'll try that

**Pull request merged**